### PR TITLE
instrument pgwire with some metrics

### PR DESCRIPTION
**Pull request by doy-materialize**
### Motivation

fixes https://github.com/MaterializeInc/cloud/issues/4146

adds some basic status metrics for slo definitions. this adds a metric for `mz_connection_status` to track network connection failures, and adds failure tracking to the existing `mz_query_total` metric.

### Tips for reviewer

this is a partial revert of #12869

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/pgwire/src/protocol.rs
Comment (jpepin) [inline src/pgwire/src/protocol.rs:414]: Any particular reason we want the client code to pass in a raw status string rather than using a constant or a dedicated function (`message_status_success.inc()` etc.)?

**Review by jpepin — APPROVED** [review]
No big blockers. Might want someone from the database side to review as well, but lgtm

File: src/pgwire/src/protocol.rs
Comment (doy-materialize) [reply src/pgwire/src/protocol.rs:414]: it seemed plausible that we might want something more detailed than just "success" and "failure" at some point in the future in some of these cases

**Review by benesch — COMMENTED** [review]
Thanks very much for this, @doy-materialize! But yes, please make sure to get a review from someone on surfaces before merging. I've requested review from @mjibson since he's the de factor owner of the pgwire crate.

File: src/pgwire/src/protocol.rs
Comment (mjibson) [reply src/pgwire/src/protocol.rs:414]: I'm not sure that using `error` here makes sense, or that using `success` for `Done` is correct. For example, `Done` is produced when there was a fatal error that will cause the connection to close (https://github.com/MaterializeInc/materialize/blob/3c9876d6bbaa2da26283f2abb6e8f90bd52192d7/src/pgwire/src/protocol.rs#L1761) and `Drain` is used when a user tries to execute a statement in an aborted transaction (which can happen trivially and doesn't mean anything bad per se).

These are not messages users ever send or receive, but are internal state machine flags. I am not currently convinced they should be measured. And if they should, probably not like this.

File: src/pgwire/src/protocol.rs
Comment (mjibson) [inline src/pgwire/src/protocol.rs:486]: This would count the number of SQL errors users type. Is that what we want?

File: src/pgwire/src/protocol.rs
Comment (mjibson) [inline src/pgwire/src/protocol.rs:472]: I have a preference that any query measuring we do happen directly on `adapter_client` (`SessionClient`) because 1) you wouldn't have to duplicate metric counting like in the current diff and 2) it'll also just work for our http sql interface.

**Review by mjibson — COMMENTED** [review]
The listed issue discusses tcp connection status and latency. I'm not sure what those mean specifically, or if they are being measured here. I'd like to understand what the expected output of this work is so I can comment on if the things we are collecting metrics on will help achieve that expectation.

For example, are we trying add metrics to be able to measure system health or user success (both? or are those actually the same thing?). I might need like a "theory of metrics and how they relate to SLAs/company goals/user goals" primer to help me understand your thoughts on this.

File: src/pgwire/src/protocol.rs
Comment (doy-materialize) [reply src/pgwire/src/protocol.rs:414]: yeah, this was just restoring a metric that previously existed in #12869, but i'm not convinced it's actually meaningful at all, i'll probably just remove it

File: src/pgwire/src/protocol.rs
Comment (doy-materialize) [reply src/pgwire/src/protocol.rs:486]: i think it's fine - in the steady state it won't be zero, but it should be fairly low, and i think we would still get signal from if this spikes unexpectedly. i don't think we should alert on it, but it could be useful signal for tracking down issues.

File: src/pgwire/src/protocol.rs
Comment (doy-materialize) [reply src/pgwire/src/protocol.rs:472]: yeah, this makes sense - i'll see what moving it in there looks like

File: src/adapter/src/coord/metrics.rs
Comment (mjibson) [inline src/adapter/src/coord/metrics.rs:19]: Does `query_total` do anything now? It seems like it's now a `sum` of `query_status`.

File: src/adapter/src/util.rs
Comment (mjibson) [inline src/adapter/src/util.rs:39]: I'm not convinced that putting this on ClientTransmitter is a good choice. `adapter_client` I think is a better fit.

File: src/adapter/src/coord/metrics.rs
Comment (doy-materialize) [reply src/adapter/src/coord/metrics.rs:19]: the only difference is that `query_total` is incremented when the query is started and `query_status` is incremented when the query finishes - i wasn't sure if people cared about that behavior. if you don't think it makes a difference either way, i could collapse them into one.

File: src/adapter/src/util.rs
Comment (doy-materialize) [reply src/adapter/src/util.rs:39]: alright, i added it in both places - happy to remove that second commit adding it to the ClientTransmitter if you think it's just unnecessary duplication

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [inline src/adapter/src/coord/command_handler.rs:245]: These should stay.

**Review by mjibson — COMMENTED** [review]
I think this PR should only be changes to pgwire. I'm working on adapter client and coord metrics elsewhere.

**Comment by doy-materialize** [comment]
@mjibson: okay, how's this?

**Review by mjibson — APPROVED** [review]

**Pull request merged**
